### ISSUE TITLE
Feature request: Import tree structure from TST on first run

### ISSUE THREAD

**Issue by eternaleye** [traxium/tabtree#71]
I have too many tabs to manually re-tree them, making migration infeasible without importing my existing trees.

In order to get past that, I used the following Perl script:

``` perl
use strict;
use warnings;

use JSON::XS;

# First four lines are file header; last is one huge JSON dump
my @lines = <>;

my $json = JSON::XS->new();

my $session = $json->decode(pop(@lines));

# Used to store a mapping from TST tab IDs to the tab JSON
my %tabids = ();

# Populate %tabids
for my $window ( @{$session->{windows}} ) {
    for my $tab ( @{$window->{tabs}} ) {
        $tabids{$tab->{extData}->{'treestyletab-id'}} = $tab;
    }
}

# Nest under parents
for my $window ( @{$session->{windows}} ) {
    for my $tab ( @{$window->{tabs}} ) {
        determine_level( $tab );
    }
}

# TST also has:
#    "treestyletab-children" - ordered list of direct child tabs, separated by '|'
#    "treestyletab-insert-after" - ID of tab on same level or less which directly precedes this
#    "treestyletab-insert-before" - like after, but reversed
# However, -after and -before are probably not needed, due to TST maintaining normal
# tab ordering as far as I can tell, and -children is simply a reflexive of -parent.

push( @lines, $json->encode($session) );

print join( '', @lines ), "\n";

sub determine_level {
     my $tab = shift;
     if ( not exists $tab->{extData}->{'treestyletab-parent'} ) {
         # Tab has no TST parent, is top-level
         return 0;
     } else {
         my $parent_id = $tab->{extData}->{'treestyletab-parent'};
         my $parent = $tabids{$parent_id};
         $tab->{extData}->{ttLevel} = 1 + determine_level( $parent );
         return $tab->{extData}->{ttLevel};
     }
}
```

I ran it on a session saved by [Session Manager](http://sessionmanager.mozdev.org/), but the same data is almost certainly maintained in memory due to SM building on the core [Session Store API](https://developer.mozilla.org/en/docs/Session_store_API).

The script is invoked as `treetab-convert.pl OLD_SESSION_FILE > NEW_SESSION_FILE`, though either deleting the old session file afterwards or altering the "name" field in the header of the new one is necessary to load the altered session.

SM keeps its sessions in `$PROFILE/sessions`.

**Comment by tomdavey** [comment]
+1. This feature would be very valuable to those migrating from TST. I too have many tabs open, in carefully structured hierarchies. It was disconcerting to install Tab Tree and find the hierarchy from TST gone. Maybe it would be good to include a warning to TST users about this, before installing Tab Tree?

**Comment by traxium** [comment]
Done. Thanks very much for the code example. The feature will be available in version 1.4.5.
I check the very first tab and If it has "treestyletab-id" then I copy the tree structure. Of course if TST was disabled and then the first tab is replaced with a newly created one and after that Tab Tree is installed then nothing will happen.
And if between disabling TST and installing Tab Tree new tabs are added then they have level 0 even if they are in the middle of a subtree. The indentation level of this tabs can be adjusted manually with Ctrl+Shift+Left/Right.

**Comment by traxium** [comment]
Ctrl+Shift+Left/Right is changed to Ctrl+Shift+,/. in #88. Nobody complains, I'll close this issue.
